pkg/querier/worker/frontend_processor.go

```diff
@@ -100,9 +99,9 @@ func (fp *frontendProcessor) processQueriesOnSingleStream(workerCtx context.Cont
 }
 
 // process loops processing requests on an established stream.
-func (fp *frontendProcessor) process(c frontendv1pb.Frontend_ProcessClient, address string, inflightQuery *atomic.Bool) (err error) {
+func (fp *frontendProcessor) process(execCtx context.Context, c frontendv1pb.Frontend_ProcessClient, address string, inflightQuery *atomic.Bool) (err error) {
 	// Build a child context so we can cancel a query when the stream is closed.
-	ctx, cancel := context.WithCancelCause(c.Context())
```

Comment:
Note, the [`grpc-go` docs mention that the `ClientStream.Context`][1]:

> ... should not be called until after Header or RecvMsg has returned. Once called, subsequent client-side retries are disabled.

I don't think this cases any issues, as this one was here for several years. That is, I consider this a "code clean-up", rather than a "fix".

[1]: https://pkg.go.dev/google.golang.org/grpc#ClientStream.Context

Reply:
does this change now mean that the frontend might cancel the stream, but we wouldn't know about it? This is the code in the frontend.

https://github.com/grafana/mimir/blob/e14de9592b4e19a69139a2d49447c789f679bf0f/pkg/frontend/v1/frontend.go#L270-L274

I'm slightly surprised no tests failed :/

Reply:
I believe it should work just fine. That is, consider the following scenario:

```
client → frontend → querier → backend
```

1. When a client cancels their request, the frontend closes its side of the stream, returning the `context.Cancel` wrapped into an RPC-error, to the querier's side of the stream. 
2. The processing loop inside the querier will `Receive` this error and bail out.
3. This will close the local context inside the `frontendProcessor.process()`, thanks to `defer cancel()`.
4. An in-flight request to the backend will be notified about the cancellation in (3).

Note, I've now looked through the tests in `querier/worker`, and added some cosmetic changes. The overall behaviour shouldn't have changed.

Reply:
makes sense, I had missed the `defer cancel()`